File: mealie/schema/event_notifications.py
from enum import Enum
from typing import Optional

from fastapi_camelcase import CamelModel


class DeclaredTypes(str, Enum):
    general = "General"
    discord = "Discord"
    gotify = "Gotify"
    pushover = "Pushover"
    home_assistant = "Home Assistant"


class EventNotificationOut(CamelModel):
    id: Optional[int]
    name: str = ""
    type: DeclaredTypes = DeclaredTypes.general
    general: bool = True
    recipe: bool = True
    backup: bool = True
    scheduled: bool = True
    migration: bool = True
    group: bool = True
    user: bool = True

    class Config:
        orm_mode = True


class EventNotificationIn(EventNotificationOut):
    notification_url: str = ""

    class Config:
        orm_mode = True


class Discord(CamelModel):
    webhook_id: str
    webhook_token: str

    @property
    def create_url(self) -> str:
        return f"discord://{self.webhook_id}/{self.webhook_token}/"


class GotifyPriority(str, Enum):
    low = "low"
    moderate = "moderate"
    normal = "normal"
    high = "high"


class Gotify(CamelModel):
    hostname: str
    token: str
    priority: GotifyPriority = GotifyPriority.normal

    @property
    def create_url(self) -> str:
        return f"gotifys://{self.hostname}/{self.token}/?priority={self.priority}"

File: frontend/api/class-interfaces/event-notifications.ts
import { BaseCRUDAPI } from "./_base";

export type EventCategory = "general" | "recipe" | "backup" | "scheduled" | "migration" | "group" | "user";
export type DeclaredTypes = "General" | "Discord" | "Gotify" | "Pushover" | "Home Assistant";
export type GotifyPriority = "low" | "moderate" | "normal" | "high";

export interface EventNotification {
  id?: number;
  name?: string;
  type?: DeclaredTypes & string;
  general?: boolean;
  recipe?: boolean;
  backup?: boolean;
  scheduled?: boolean;
  migration?: boolean;
  group?: boolean;
  user?: boolean;
}

export interface CreateEventNotification extends EventNotification {
  notificationUrl?: string;
}

const prefix = "/api";

const routes = {
  aboutEventsNotifications: `${prefix}/about/events/notifications`,
  aboutEventsNotificationsTest: `${prefix}/about/events/notifications/test`,

  aboutEventsNotificationsId: (id: number) => `${prefix}/about/events/notifications/${id}`,
};

export class NotificationsAPI extends BaseCRUDAPI<EventNotification, CreateEventNotification> {
  baseRoute = routes.aboutEventsNotifications;
  itemRoute = routes.aboutEventsNotificationsId;
  /** Returns the Group Data for the Current User
   */
  async testNotification(id: number | null = null, testUrl: string | null = null) {
    return await this.requests.post(routes.aboutEventsNotificationsTest, { id, testUrl });
  }
}
